### DISCUSSION TITLE
How to do the equivalent of `jest.createMockFromModule('fs');` in a `__mocks__/fs.js` file?

### DISCUSSION THREAD

**Discussion by garyking** [vitest-dev/vitest#5091] [Q&A]
I have a project with Jest that I'm trying to convert to Vitest. There are many files in `__mocks__`, like `fs.js`, which contains just one line: `module.exports = jest.createMockFromModule('fs');`

The intention was to make it so that whenever `fs` was called in a test, it wouldn't actually run anything, since every method in the module was mocked. Also, the other advantage of this was that I didn't have to use `jest.mock('fs')` in any files; `fs` was automatically mocked, using the `__mocks__/fs.js` file.

Is there any way to do this same thing in Vitest? Is it something like this?:

```ts
import { vi } from 'vitest'

const fs = await vi.importMock('fs')

export default fs
```

And then in my test file, I would have to have `vi.mock('fs')` in it? Is there a way to make it so that I don't need to specify `vi.mock('fs')` in every test file? The nice thing about the Jest way, was that it was a safeguard against accidentally running a damaging `fs` method.

Also, I notice that `importMock` doesn't seem to actually create mocked versions of each method; I had to separately add `fs.readFileSync = vi.fn();` in order to turn it into a mocked function. Or am I doing something wrong? Is there a way for Vitest to automatically create mocked versions of each method, like in Jest?

Thanks in advance!

**Comment by sheremet-va** [comment]
`importMock` returns a Promise, so you need to await it. Your example also doesn't cover named exports, only default ones.

To auromock fs in every test, you can call `vi.mock("fs")` in a setup file.

**Reply by garyking** [reply]
In my code, I use `await`, I just forgot to add it in the example above; I've added it now. 

How can I mock named exports, rather than just the default export? Is there a way to mock them all at once, or do I actually have to mock each one individually?

Thanks in advance!

**Reply by hi-ogawa** [reply]
You might not need either `__mocks__/fs.js` nor `vi.importMock("fs")`. Instead, have you tried calling `vi.mock("fs")` in `setupFiles`? https://vitest.dev/config/#setupfiles.

> the other advantage of this was that I didn't have to use `jest.mock('fs')` in any files; fs was automatically mocked, using the `__mocks__/fs.js file`.

Btw, Jest doc says they require calling `jest.mock('fs')` explicitly even if there is `__mocks__/fs.js` https://jestjs.io/docs/manual-mocks#mocking-node-modules So, maybe did you have `jest.mock('fs')` in setup files before?

**Reply by garyking** [reply]
Thanks, using the setup files seems to help!
